From 5ac397f2bf496a0e8ee34cf737be22065ffb92e1 Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Fri, 17 Jul 2026 11:11:01 +0200
Subject: [PATCH] src: do not merge a set with a erroneous one

JIRA: https://issues.redhat.com/browse/RHEL-190549
Upstream Status: nftables commit ea011231c06cbe828cf6056bc9c3d116e1f528d5

commit ea011231c06cbe828cf6056bc9c3d116e1f528d5
Author: Florian Westphal <fw@strlen.de>
Date:   Fri Jan 12 13:19:26 2024 +0100

    src: do not merge a set with a erroneous one

    The included sample causes a crash because we attempt to
    range-merge a prefix expression with a symbolic expression.

    The first set is evaluated, the symbol expression evaluation fails
    and nft queues an error message ("Could not resolve hostname").

    However, nft continues evaluation.

    nft then encounters the same set definition again and merges the
    new content with the preceeding one.

    But the first set structure is dodgy, it still contains the
    unresolved symbolic expression.

    That then makes nft crash (assert) in the set internals.

    There are various different incarnations of this issue, but the low
    level set processing code does not allow for any partially transformed
    expressions to still remain.

    Before:
    nft --check -f tests/shell/testcases/bogons/nft-f/invalid_range_expr_type_binop
    BUG: invalid range expression type binop
    nft: src/expression.c:1479: range_expr_value_low: Assertion `0' failed.

    After:
    nft --check -f tests/shell/testcases/bogons/nft-f/invalid_range_expr_type_binop
    invalid_range_expr_type_binop:4:18-25: Error: Could not resolve hostname: Name or service not known
    elements = { 1&.141.0.1 - 192.168.0.2}
                 ^^^^^^^^

    Signed-off-by: Florian Westphal <fw@strlen.de>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Phil Sutter <psutter@redhat.com>
---
 include/rule.h                                       |  2 ++
 src/evaluate.c                                       |  4 +++-
 src/intervals.c                                      |  2 +-
 .../bogons/nft-f/invalid_range_expr_type_binop       | 12 ++++++++++++
 4 files changed, 18 insertions(+), 2 deletions(-)
 create mode 100644 tests/shell/testcases/bogons/nft-f/invalid_range_expr_type_binop

diff --git a/include/rule.h b/include/rule.h
index a82c794..fe281cc 100644
--- a/include/rule.h
+++ b/include/rule.h
@@ -331,6 +331,7 @@ void rule_stmt_insert_at(struct rule *rule, struct stmt *nstmt,
  * @policy:	set mechanism policy
  * @automerge:	merge adjacents and overlapping elements, if possible
  * @comment:	comment
+ * @errors:	expr evaluation errors seen
  * @desc.size:		count of set elements
  * @desc.field_len:	length of single concatenated fields, bytes
  * @desc.field_count:	count of concatenated fields
@@ -355,6 +356,7 @@ struct set {
 	bool			root;
 	bool			automerge;
 	bool			key_typeof_valid;
+	bool			errors;
 	const char		*comment;
 	struct {
 		uint32_t	size;
diff --git a/src/evaluate.c b/src/evaluate.c
index a294bbd..256ca10 100644
--- a/src/evaluate.c
+++ b/src/evaluate.c
@@ -4718,8 +4718,10 @@ static int elems_evaluate(struct eval_ctx *ctx, struct set *set)
 
 		__expr_set_context(&ctx->ectx, set->key->dtype,
 				   set->key->byteorder, set->key->len, 0);
-		if (expr_evaluate(ctx, &set->init) < 0)
+		if (expr_evaluate(ctx, &set->init) < 0) {
+			set->errors = true;
 			return -1;
+		}
 		if (set->init->etype != EXPR_SET)
 			return expr_error(ctx->msgs, set->init, "Set %s: Unexpected initial type %s, missing { }?",
 					  set->handle.set.name, expr_name(set->init));
diff --git a/src/intervals.c b/src/intervals.c
index 0618b49..31adb82 100644
--- a/src/intervals.c
+++ b/src/intervals.c
@@ -136,7 +136,7 @@ static void set_sort_splice(struct expr *init, struct set *set)
 	set_to_range(init);
 	list_expr_sort(&init->expressions);
 
-	if (!existing_set)
+	if (!existing_set || existing_set->errors)
 		return;
 
 	if (existing_set->init) {
diff --git a/tests/shell/testcases/bogons/nft-f/invalid_range_expr_type_binop b/tests/shell/testcases/bogons/nft-f/invalid_range_expr_type_binop
new file mode 100644
index 0000000..514d6ff
--- /dev/null
+++ b/tests/shell/testcases/bogons/nft-f/invalid_range_expr_type_binop
@@ -0,0 +1,12 @@
+table ip x {
+	map z {
+		type ipv4_addr : ipv4_addr
+		elements = { 1&.141.0.1 - 192.168.0.2}
+	}
+
+	map z {
+		type ipv4_addr : ipv4_addr
+		flags interval
+		elements = { 10.141.0.0, * : 192.168.0.4 }
+	}
+}
